src/errors.rs

```diff
@@ -1,50 +1,62 @@
-use super::Box;
+use super::{dealloc, Channel};
 use core::fmt;
 use core::mem;
+use core::ptr::NonNull;
 
 /// An error returned when trying to send on a closed channel. Returned from
 /// [`Sender::send`] if the corresponding [`Receiver`] has already been dropped.
 ///
 /// The message that could not be sent can be retreived again with [`SendError::into_inner`].
 pub struct SendError<T> {
-    channel_ptr: *mut super::Channel<T>,
+    channel_ptr: NonNull<Channel<T>>,
 }
 
 unsafe impl<T: Send> Send for SendError<T> {}
+unsafe impl<T: Sync> Sync for SendError<T> {}
 
 impl<T> SendError<T> {
-    pub(crate) const fn new(channel_ptr: *mut super::Channel<T>) -> Self {
+    /// Safety: by calling this function, the caller semantically transfers ownership of the
+    /// channel's resources to the created `SendError`. Thus the caller must ensure that the
+    /// pointer is not used in a way which would violate this ownership transfer.
+    pub(crate) const unsafe fn new(channel_ptr: NonNull<Channel<T>>) -> Self {
         Self { channel_ptr }
     }
 
     /// Consumes the error and returns the message that failed to be sent.
     #[inline]
     pub fn into_inner(self) -> T {
-        // SAFETY: The reference won't be used after it is freed in this method
-        let channel: &mut super::Channel<T> = unsafe { &mut *self.channel_ptr };
+        let channel_ptr = self.channel_ptr;
+        // SAFETY: we have ownership of the channel
+        let channel: &Channel<T> = unsafe { channel_ptr.as_ref() };
```

Comment:
This is a common pattern now in the code with your change. To be a bit nitpicky I think `let channel` should be created after `forget(self)`. Since this statement is no longer depending on `self` it has no technical reason to be before it. And it's good practice to declare variables as close to where they are used as possible.

As I said, it's a bit nitpicky, but I think the following order would be cleaner:

```rust
let channel_ptr = self.channel_ptr;
mem::forget(self);

let channel: &Channel<T> = unsafe { channel_ptr.as_ref() };
// Do stuff with `channel`
```

Reply:
Resolved!